src/ansiblelint/rules/no_jinja_when.py

```diff
@@ -66,6 +69,40 @@ def matchtask(
     ) -> bool | str:
         return "when" in task.raw_task and not self._is_valid(task.raw_task["when"])
 
+    def transform(
+        self,
+        match: MatchError,
+        lintable: Lintable,
+        data: CommentedMap | CommentedSeq | str,
+    ) -> None:
+        if match.tag == self.id:
+            task = self.seek(match.yaml_path, data)
+            key_to_check = ("when", "changed_when", "failed_when")
+            for _ in range(len(task)):
+                k, v = task.popitem(False)
+                if k == "roles" and isinstance(v, list):
+                    transform_for_roles(v, key_to_check=key_to_check)
+                elif k in key_to_check:
+                    processed_value = re.sub(r"{{ (.*?) }}", r"\1", v)
+                    v = processed_value
+                task[k] = v
+            match.fixed = True
+
+
+def transform_for_roles(v: list[Any], key_to_check: tuple[str, ...]) -> None:
+    """Additional transform logic in case of roles."""
+    for idx, new_dict in enumerate(v):
+        for new_key, new_value in new_dict.items():
+            if new_key in key_to_check:
+                if isinstance(new_value, list):
+                    for index, nested_value in enumerate(new_value):
+                        processed_value = re.sub(r"{{ (.*?) }}", r"\1", nested_value)
```

Comment:
here also, assigned it directly to `new_value[index]`, so no need of another variable